include/mra/kernels/multiply.h

```diff
@@ -11,64 +11,96 @@ namespace mra {
   template<mra::Dimension NDIM>
   SCOPE size_type multiply_tmp_size(size_type K) {
     const size_type K2NDIM = std::pow(K,NDIM);
-    return 4*K2NDIM; // workspace, r, r1, and r2
+    return 25*K2NDIM; // workspace, r1*8, cnodeA*8, cnodeB*8,
   }
 
   namespace detail {
 
     template <typename T, Dimension NDIM>
     DEVSCOPE void multiply_kernel_impl(
       const Domain<NDIM>& D,
+      const Key<NDIM>& keyA,
+      const Key<NDIM>& keyB,
       const TensorView<T, NDIM>& nodeA,
       const TensorView<T, NDIM>& nodeB,
+      TensorView<T, NDIM+1>& cnodeA,
+      TensorView<T, NDIM+1>& cnodeB,
       TensorView<T, NDIM>& nodeR,
-      TensorView<T, NDIM>& r,
-      TensorView<T, NDIM>& r1,
-      TensorView<T, NDIM>& r2,
+      TensorView<T, NDIM+1>& r1,
       T* workspace,
+      const TensorView<T, 2>& phi,
       const TensorView<T, 2>& phiT,
       const TensorView<T, 2>& phibar,
-      Key<NDIM> key,
+      const TensorView<T, 1>& quad_x,
       size_type K)
     {
-      // convert coeffs to function values
-      transform(nodeA, phiT, r1, workspace);
-      transform(nodeB, phiT, r2, workspace);
-      const T scale = std::pow(T(2), T(0.5 * NDIM * key.level())) / std::sqrt(D.template get_volume<T>());
+      if (keyA.level() > keyB.level()){
+        T scale;
+        for (int i=0; i< keyA.num_children(); ++i){
+          fcube_for_mul(D, keyA.child_at(i), keyB, nodeB, cnodeB[i], phibar, phi, quad_x, K, workspace);
+          scale = std::sqrt(D.template get_volume<T>()*std::pow(T(0.5), T(NDIM*keyB.right_child().level())));
+          cnodeB[i] *= scale;
 
-      foreach_idx(nodeA, [&](size_type i) {
-          r[i] = scale * r1[i] * r2[i];
+          fcube_for_mul(D, keyA.child_at(i), keyA, nodeA, cnodeA[i], phibar, phi, quad_x, K, workspace);
+          scale = std::sqrt(D.template get_volume<T>()*std::pow(T(0.5), T(NDIM*keyA.right_child().level())));
+          cnodeA[i] *= scale;
+        }
+      }
+      else if (keyA.level() <= keyB.level()){
+        T scale;
+        for (int i=0; i< keyA.num_children(); ++i){
+          fcube_for_mul(D, keyB.child_at(i), keyA, nodeA, cnodeA[i], phibar, phi, quad_x, K, workspace);
+          scale = std::sqrt(D.template get_volume<T>()*std::pow(T(0.5), T(NDIM*keyB.right_child().level())));
+          cnodeB[i] *= scale;
+
+          fcube_for_mul(D, keyB.child_at(i), keyB, nodeB, cnodeB[i], phibar, phi, quad_x, K, workspace);
+          scale = std::sqrt(D.template get_volume<T>()*std::pow(T(0.5), T(NDIM*keyA.right_child().level())));
+          cnodeA[i] *= scale;
+        }
+      }
+
+      // fcube_for_mul() returns function values evaluated at quadrature points
+      for (int i=0; i< keyA.num_children(); ++i)
+        foreach_idx(cnodeA[i], [&](size_type j) {
+          cnodeA[i][j] = cnodeA[i][j] * cnodeB[i][j];
       });
```

Comment:
I don't think this will compile. But we could simplify this to

```suggestion
      foreach_idx(cnodeA, [&](size_type i) {
        cnodeA[i] = cnodeA[i] * cnodeB[i];
      });
```

to do point-wise multiplication over all *all* elements across all children.

Reply:
`cnodeA` is `NDIM+1`. If the ordering of child nodes is going to be the same, then this would work I believe. It does compile though!